lib/plugins/aws/package/compile/events/apiGateway/index.js

```diff
@@ -20,15 +21,261 @@ const getMethodAuthorization = require('./lib/method/authorization');
+    },
+    path: { type: 'string' },
+    operationId: { type: 'string' },
+    async: { type: 'boolean' },
+    private: { type: 'boolean' },
+    request: requestSchema,
+    response: responseSchema,
+    authorizer: authorizerSchema,
+  },
+  required: ['path'],
+  additionalProperties: false,
+};
+
+const lambdaProxyIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: {
+      enum: extendWithUppercase(['lambda_proxy']),
+    },
+  },
+  required: ['integration'],
+});
+
+const mockIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: {
+      enum: extendWithUppercase(['mock']),
+    },
+  },
+  required: ['integration'],
+});
+
+const awsIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: {
+      enum: extendWithUppercase(['aws', 'aws_proxy']),
+    },
+  },
+  required: ['integration'],
+});
+
+const httpIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: {
+      enum: extendWithUppercase(['http', 'http_proxy']),
+    },
+    request: {
+      required: ['uri'],
+    },
+  },
+  required: ['integration'],
+});
+
+const httpVpcLinkIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    connectionType: { enum: ['vpc-link', 'VPC_LINK'] },
+    connectionId: { type: 'string' },
+    integration: {
+      enum: extendWithUppercase(['http', 'http_proxy']),
+    },
+    request: {
+      required: ['uri'],
+    },
+  },
+  required: ['integration', 'connectionType', 'connectionId'],
+});
+
+const lambdaIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: {
+      enum: extendWithUppercase(['lambda']),
+    },
+    authorizer: {
+      properties: {
+        claims: { type: 'array', items: { type: 'string' } },
+      },
+    },
+  },
+  required: ['integration'],
+});
+
+const cognitoIntegrationSchema = _.merge({}, baseIntegrationSchema, {
+  properties: {
+    integration: {
+      enum: extendWithUppercase(['cognito_user_pools']),
+    },
+    authorizer: {
+      required: ['name'],
+    },
+  },
+  required: ['integration'],
+});
+
 class AwsCompileApigEvents {
   constructor(serverless, options) {
     this.serverless = serverless;
     this.options = options;
     this.provider = this.serverless.getProvider('aws');
 
     this.serverless.configSchemaHandler.defineFunctionEvent('aws', 'http', {
-      // TODO: Complete schema, see https://github.com/serverless/serverless/issues/8018
-      anyOf: [{ type: 'string' }, { type: 'object' }],
+      oneOf: [
+        { type: 'string', pattern: `^(${httpMethods.join('|')}) \\S+` },
+        baseIntegrationSchema,
+        awsIntegrationSchema,
+        cognitoIntegrationSchema,
+        httpIntegrationSchema,
+        httpVpcLinkIntegrationSchema,
+        lambdaIntegrationSchema,
+        lambdaProxyIntegrationSchema,
+        mockIntegrationSchema,
```

Comment:
I believe we expect `method` and `path` unconditionally, while above configuration, seem to expect different set of properties exclusively (and most of them do not require `method` and `path`), that doesn't look right.

In my understanding, schema for `http` is either _string_ (in form of `method path`), or _single_ version of object where `method` and `path` are required.

In similar manner as in `httpApi` event here: https://github.com/serverless/serverless/blob/5851bcadcf6016914013d447de899cc98efb8346/lib/plugins/aws/package/compile/events/httpApi/index.js#L57-L79

but ofc we will have more properties in case of this event

Reply:
I declared `path` as required in `baseIntegrationSchema`, so all other integration schemas inherit that.
I did fail to declare `method` as required as well.
The reason I used multiple object schemas instead of one is there many conditions regarding integration types. That is what I referred to when I [wrote about if/else](https://github.com/serverless/serverless/pull/8301#issuecomment-702242407) - if I go with one object for all integration types, then I would have to use many if/else clauses to declare all constraints.

Reply:
Indeed, sorry I failed to see, that other objects, share all properties from `baseIntegrationSchema`